File: cpp/src/v0/tick_source.cpp
#include "tick_source.hpp"

#include <fstream>
#include <stdexcept>

namespace ytta::v0 {
namespace {

std::optional<std::string_view> json_string_field(std::string_view line,
                                                  std::string_view key) {
  const std::string needle = "\"" + std::string(key) + "\":\"";
  const auto pos = line.find(needle);
  if (pos == std::string_view::npos) {
    return std::nullopt;
  }
  const auto start = pos + needle.size();
  const auto end = line.find('"', start);
  if (end == std::string_view::npos) {
    return std::nullopt;
  }
  return line.substr(start, end - start);
}

std::optional<std::int64_t> json_int_field(std::string_view line,
                                           std::string_view key) {
  const std::string needle = "\"" + std::string(key) + "\":";
  const auto pos = line.find(needle);
  if (pos == std::string_view::npos) {
    return std::nullopt;
  }
  auto start = pos + needle.size();
  while (start < line.size() &&
         (line[start] == ' ' || line[start] == '\t')) {
    ++start;
  }
  std::size_t end = start;
  if (end < line.size() && (line[end] == '-' || line[end] == '+')) {
    ++end;
  }
  while (end < line.size() && line[end] >= '0' && line[end] <= '9') {
    ++end;
  }
  if (end == start) {
    return std::nullopt;
  }
  return std::stoll(std::string(line.substr(start, end - start)));
}

}  // namespace

std::optional<Tick> parse_tick_line(std::string_view line) {
  if (line.empty() || line[0] == '#') {
    return std::nullopt;
  }
  const auto type = json_string_field(line, "type");
  if (!type || *type != "tick") {
    return std::nullopt;
  }
  Tick t;
  const auto ts = json_int_field(line, "ts_ns");
  const auto price = json_int_field(line, "price");
  const auto qty = json_int_field(line, "qty");
  const auto seq = json_int_field(line, "seq");
  const auto instrument = json_string_field(line, "instrument");
  const auto side = json_string_field(line, "side");
  if (!ts || !price || !qty || !seq || !instrument || !side) {
    throw std::runtime_error("malformed tick line");
  }
  t.ts_ns = *ts;
  t.price = *price;
  t.qty = *qty;
  t.seq = static_cast<std::uint64_t>(*seq);
  t.instrument = std::string(*instrument);
  if (*side == "BUY") {
    t.side = Side::Buy;
  } else if (*side == "SELL") {
    t.side = Side::Sell;
  } else {
    throw std::runtime_error("invalid tick side");
  }
  return t;
}

std::vector<Tick> TickSource::load() const {
  std::ifstream in(path_);
  if (!in) {
    throw std::runtime_error("failed to open fixture: " + path_);
  }
  std::vector<Tick> ticks;
  std::string line;
  while (std::getline(in, line)) {
    if (!line.empty() && line.back() == '\r') {
      line.pop_back();
    }
    if (auto t = parse_tick_line(line)) {
      ticks.push_back(std::move(*t));
    }
  }
  return ticks;
}

}  // namespace ytta::v0

File: rust/src/v0/tick_source.rs
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::path::Path;

use super::types::{Side, Tick};

fn json_string_field(line: &str, key: &str) -> Option<String> {
    let needle = format!("\"{key}\":\"");
    let start = line.find(&needle)? + needle.len();
    let end = line[start..].find('"')? + start;
    Some(line[start..end].to_string())
}

fn json_int_field(line: &str, key: &str) -> Option<i64> {
    let needle = format!("\"{key}\":");
    let start = line.find(&needle)? + needle.len();
    let rest = line[start..].trim_start();
    let end = rest
        .find(|c: char| !(c.is_ascii_digit() || c == '-' || c == '+'))
        .unwrap_or(rest.len());
    rest[..end].parse().ok()
}

pub fn parse_tick_line(line: &str) -> Option<Result<Tick, String>> {
    let line = line.trim();
    if line.is_empty() || line.starts_with('#') {
        return None;
    }
    let ty = json_string_field(line, "type")?;
    if ty != "tick" {
        return None;
    }
    Some((|| {
        let ts_ns = json_int_field(line, "ts_ns").ok_or("missing ts_ns")?;
        let price = json_int_field(line, "price").ok_or("missing price")?;
        let qty = json_int_field(line, "qty").ok_or("missing qty")?;
        let seq = json_int_field(line, "seq").ok_or("missing seq")? as u64;
        let instrument = json_string_field(line, "instrument").ok_or("missing instrument")?;
        let side_s = json_string_field(line, "side").ok_or("missing side")?;
        let side = match side_s.as_str() {
            "BUY" => Side::Buy,
            "SELL" => Side::Sell,
            other => return Err(format!("invalid side {other}")),
        };
        Ok(Tick {
            ts_ns,
            instrument,
            side,
            price,
            qty,
            seq,
        })
    })())
}

pub fn load_ticks(path: impl AsRef<Path>) -> Result<Vec<Tick>, String> {
    let file = File::open(path.as_ref()).map_err(|e| e.to_string())?;
    let reader = BufReader::new(file);
    let mut ticks = Vec::new();
    for (i, line) in reader.lines().enumerate() {
        let line = line.map_err(|e| e.to_string())?;
        if let Some(parsed) = parse_tick_line(&line) {
            ticks.push(parsed.map_err(|e| format!("line {}: {e}", i + 1))?);
        }
    }
    Ok(ticks)
}
